contracts/utils/SlotDerivation.sol

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: MIT
+// This file was procedurally generated from scripts/generate/templates/SlotDerivation.js.
+
+pragma solidity ^0.8.20;
+
+/**
+ * @dev Library for computing storage (and transient storage) locations from namespaces and deriving slots
+ * corresponding to standard patterns. The derivation method for array and mapping matches the storage layout used by
+ * the solidity language / compiler.
+ *
+ * See https://docs.soliditylang.org/en/v0.8.20/internals/layout_in_storage.html#mappings-and-dynamic-arrays[Solidity docs for mappings and dynamic arrays.].
+ *
+ * Example usage:
+ * ```solidity
+ * contract Example {
+ *     // Add the library methods
+ *     using SlotDerivation for bytes32;
+ *
+ *     // Declare a namespace
+ *     string private constant _NAMESPACE = "<namespace>" // eg. OpenZeppelin.Slot
+ *
+ *     function storagePointer() internal view returns (bytes32) {
+ *         return _NAMESPACE.erc7201Slot(); // or erc1967Slot()
+ *     }
+ * }
+ * ```
+ *
+ * TIP: Consider using this library along with {StorageSlot}.
+ */
+library SlotDerivation {
+    /**
+     * @dev Derive an ERC-1967 slot from a string (namespace).
+     */
+    function erc1967Slot(string memory namespace) internal pure returns (bytes32 slot) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            slot := sub(keccak256(add(namespace, 0x20), mload(namespace)), 1)
+        }
+    }
```

Comment:
For ERC1967 there are only 3 valid namespaces:

- 'eip1967.proxy.implementation'
- 'eip1967.proxy.admin'
- 'eip1967.proxy.beacon'

I think we should not expose this function since it may be dangerous (i.e. is it collision resistant as erc7201 is?), and rather only expose the slots that are already defined in the EIP. Our go-to recommendation for deriving slots should be always ERC7201

Reply:
Concretely, we might do:

```solidity
function erc1967Admin(string memory namespace) internal pure returns (bytes32 slot) {
    return 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
}

function erc1967Implementation(string memory namespace) internal pure returns (bytes32 slot) {
    return 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
}

function erc1967Beacon(string memory namespace) internal pure returns (bytes32 slot) {
    return 0xa3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50;
}
```

Reply:
I'm more on the side of @ernestognw on this one. It's true that the formula is used in other places, but I don't know that I would call those "ERC-1967".

Reply:
So I guess we remove this function then